src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhereKeywordRecommender.cs

```diff
@@ -109,13 +109,17 @@ private static bool IsTypeParameterConstraintContext(CSharpSyntaxContext context
                 }
             }
 
-            // class C<T> where T : IGoo |
-            // delegate void D<T> where T : IGoo |
-            var constraintClause = token.GetAncestor<TypeParameterConstraintClauseSyntax>();
-
-            if (constraintClause != null)
+            // Check constraints only if target token is the last in the TypeParameterConstraintSyntax node.
+            // Otherwise we can be in incomplete constraint and "where" keyword will not be valid suggestion.
+            // See https://github.com/dotnet/roslyn/issues/30785 for example
+            if (token.IsLastTokenOfNode<TypeParameterConstraintSyntax>())
             {
-                if (constraintClause.Constraints.Any(c => token == c.GetLastToken(includeSkipped: true)))
+                // class C<T> where T : IGoo |
+                // delegate void D<T> where T : IGoo |
+                var constraintClause = token.GetAncestor<TypeParameterConstraintClauseSyntax>();
+
+                if (constraintClause is not null &&
+                    constraintClause.Constraints.Any(c => token == c.GetLastToken(includeSkipped: true)))
```

Comment:
```suggestion
                var constraintClause = token.GetAncestor<TypeParameterConstraintClauseSyntax>();

                if (constraintClause is not null &&
                    constraintClause.Constraints.Any(c => token == c.GetLastToken(includeSkipped: true)))
```

can't all of thise be dleeted now?

Reply:
Indeed!